### PR TITLE
GH-1847: Add MIT Restaurant NER Dataset

### PR DESCRIPTION
**Pull request by megantosh**
I took this assignment as a chance to test how Flair interacts with some sentences in Arabic and summed it up in this Jupyter Notebook:  [ExposéTutorial.ipynb.zip](https://github.com/flairNLP/flair/files/5171526/ExposeTutorial.ipynb.zip)

The standard tutorial code has been hence slightly modified. Some inline comments start with 'MM:'
A more elaborate discussion on that via Email. 

Modified Dataset attached in #1847 
test the task via: [FlairTask.py.zip](https://github.com/flairNLP/flair/files/5171498/FlairTask.py.zip)

-----
*Task Description as received on 20200901*

# Task: Train Named Entity Recognition Model for Restaurant Reviews and add it to Flair

In this task, you will use the Flair library to train a simple sequence labeling model to perform named entity recognition (NER). In NER, we want to find the names of certain entities (such as people, organizations, or genes) in text.  

# To complete this task, do the following steps: 

- Complete Flair tutorial 1 and tutorial 2 to familiarize yourself with the basic concepts of the library. In particular, try out some taggers like the included NER tagger on some text and look at the tagging results.  

- Read tutorial 3 to get an idea of different word embeddings in Flair. Look at standard Word Embeddings, Flair Embeddings and TransformerWordEmbeddings (BERT). Understand how StackedEmbeddings work.

- Now go through tutorial 6 to understand how to load a dataset in Flair (and especially how sequence labeling datasets are formatted). Then train a sequence labeling model like shown in tutorial 7 using one of the datasets in Flair (for instance UD_ENGLISH)

- Tag some text with the model you’ve created and look at the results

- Now train another model, this time with a dataset that is not included in Flair. Use this dataset of restaurant reviews tagged with information like which word is a “dish” and which a “cuisine”. 

- Download this dataset to your machine. Load this dataset as shown in tutorial 6 here. It is possible that the dataset is not in column format, if so, first reformat it. One you’ve loaded the dataset, train a new model like you did before.

- Tag some text with the model you’ve created and look at the results 

- Now add support for this dataset to Flair. The goal is to have an object like UD_ENGLISH that automatically downloads the dataset to your machine when it is first called and loads the data in the correct format. Do this in the following way: 

     - First, clone the Flair repository to your machine (using “git clone”)

     - Then, go into the file flair/datasets/sequence_labeling.py - this is where all sequence labeling datasets lie

     - Look at the object CONLL_03_DUTCH. This object downloads the files for this corpus and initializes a ColumnCorpus with the right column_format. Make a copy of this object and rename it to MITRestaurant. 

     - Change the object so that it downloads and loads your dataset. 

     - One this works, do a Pull Request (PR) to Flair. You can find lots of information online on how to do a PR.

----

### CODE REVIEW

File: flair/datasets/sequence_labeling.py [hunk 1]
```diff
@@ -447,6 +447,53 @@ def __init__(
...
+        dataset_name = self.__class__.__name__.lower()
+
+        # default dataset folder is the cache root
+        if not base_path:
+            base_path = Path(flair.cache_root) / "datasets"
+        data_folder = base_path / dataset_name
+
+        # download data if necessary
+        mit_restaurants_path = "https://megantosh.s3.eu-central-1.amazonaws.com/MITRestoCorpus/"
+        cached_path(f"{mit_restaurants_path}dev.txt", Path("datasets") / dataset_name)
```
Comment (alanakbik) [inline flair/datasets/sequence_labeling.py:484]: There is no `dev.txt` in the MIT restaurants dataset. This downloads an empty file.

File: flair/datasets/sequence_labeling.py [hunk 2]
```diff
@@ -447,6 +447,53 @@ def __init__(
             document_separator_token=None if not document_as_sequence else "-DOCSTART-",
         )
 
+class MIT_Restaurants(ColumnCorpus):
```
Comment (alanakbik) [inline flair/datasets/sequence_labeling.py:450]: Can you capitalize this to `MIT_RESTAURANTS`? We currently capitalize all prepared corpora.

**Review by alanakbik — CHANGES_REQUESTED** [review]
Looks good, just one problem: If I run 

```python
corpus = MIT_Restaurants()
print(corpus)
```

it prints: 

~~~
Corpus: 7660 train + 0 dev + 1521 test sentences
~~~

The problem is that no `dev.txt` exists, but the download routine somehow creates an empty file. To fix this, remove the download attempt (the `cached_path` call) and delete the erroneously downloaded `dev.txt` on your local machine. When you then run the corpus again, it should only have `train.txt` and `test.txt` in the dataset folder. In this case, the corpus object will automatically sample a dev part form the train part.

File: flair/datasets/sequence_labeling.py (hunk 2 above)
Comment (megantosh) [reply flair/datasets/sequence_labeling.py]: Done.
[New test file - FlairTask.py](https://megantosh.s3.eu-central-1.amazonaws.com/MITRestoCorpus/FlairTask.py)

File: flair/datasets/sequence_labeling.py (hunk 1 above)
Comment (megantosh) [reply flair/datasets/sequence_labeling.py]: Ah great! I was not aware Flair does  the sampling. Here is the new
Corpus size: 
 Training split: 6894 
 Test split: 1521 
 Dev split:766
as they show locally and from the downloaded corpus

**Review by alanakbik — APPROVED** [review]

**Comment by alanakbik** [comment]
@megantosh thanks for adding this!

**Pull request merged**